Duplicati/Library/Main/BackendManager.cs

```diff
@@ -1212,6 +1212,11 @@ public void Put(VolumeWriterBase item, IndexVolumeWriter indexfile = null, bool
                 req2 = new FileEntryItem(OperationType.Put, indexfile.RemoteFilename);
                 req2.LocalTempfile = indexfile.TempFile;
                 req.Indexfile = new Tuple<IndexVolumeWriter, FileEntryItem>(indexfile, req2);
+
+                indexfile.FinishVolume(req.Hash, req.Size);
+                indexVolumeFinishedCallback?.Invoke();

```

Comment:
Is it true that if `indexfile` is not null, then we should _always_ want to write the blocklists as defined [here](https://github.com/duplicati/duplicati/commit/04662bace1393b428e39cdc89a276ca38b44b618#diff-59539f1832ed42c3579b81f766e18bc0R301)?  I'm wondering if the method signature (like before) still invites the possibility to provide a non-null `indexfile` without the callback to write the blocklists.

It would be nice if we could inline the body of [indexVolumeFinished](https://github.com/duplicati/duplicati/commit/04662bace1393b428e39cdc89a276ca38b44b618#diff-59539f1832ed42c3579b81f766e18bc0R301) here, but it seems that we may be missing an instance to a `LocalDatabase`.

Reply:
If I am understanding your question correctly, I think in the case of the `CompactHandler` that yes we do always want to write the blocklists since blocks have just been moved to a new volume. If there are no blocklists then the foreach that calls `db.GetBlocklists` will just return an empty list and nothing will happen.
Yes a new call could be made to `Put` and they don't provide the callback, but they would have to be in the context of needing to do something so I would hope testing would show that they didn't provide the callback if needed.

For everyone else that calls `Put` they could use the callback function to do something else, they don't have to write blocklists, or they can ignore the callback completely by not passing one in.
This `Put` function is used by the `CompactHandler`, `PurgeFilesHandler`, and the `RepairHandler`. The `RepairHandler` does create an [index file](https://github.com/duplicati/duplicati/blob/837a787cbf233c07f739e31de3025ebbf94f1229/Duplicati/Library/Main/Operation/RepairHandler.cs#L312) and calls `Put` but it uses the `VolumeWriterBase item` parameter to upload the file and does not use the `indexfile` parameter.

I agree, since the necessary code is really just a foreach with a function call it would be nice to inline here. (Really it would be nice to have it centralized in one place and have it done before calling `Put`, but that is obviously way more involved.)

I'm not sure I actually answered your question, and then I rambled on.

Reply:
That makes sense.  This might be a case where having different overloads as opposed to one method with optional parameters could help clarify the code.  But we can always refactor this later as needed.